### PR TITLE
[NFC] Refactor transform dialect matcher.

### PR DESCRIPTION
**Pull request by ThomasRaoux**
Move the matcher closer to the transform dialect creation to bypass the attributes. Work in progress.

### CODE REVIEW

**Review by nicolasvasilache — APPROVED** [review]
LGTM, nit: mark NFC

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectJitterPass.cpp [hunk 1]
```diff
@@ -396,28 +111,15 @@ class TransformDialectJitterPass
...
-    b.setInsertionPointToStart(&topLevelTransformRegion.front());
-    b.create<::transform_ext::CanonicalizedSequenceOp>(
-        target->getLoc(), transform::FailurePropagationMode::Suppress,
-        [](OpBuilder &b, Location loc, Value variantH) {
-          ImplicitLocOpBuilder ib(loc, b);
-          buildReductionCudaStrategy(ib, variantH);
-          b.create<transform::YieldOp>(loc);
+    Region *topLevelTransformRegion = nullptr;
+    target->walk(
+        [&](::transform_ext::CanonicalizedSequenceOp transformTopLevelOp) {
```
Comment (ftynse) [inline compiler/src/iree/compiler/Codegen/Common/TransformDialectJitterPass.cpp:116]: A more robust way to do this: find the op with `PossibleTopLevelTransformOpTrait` that has no operands.

Comment (ThomasRaoux) [reply compiler/src/iree/compiler/Codegen/Common/TransformDialectJitterPass.cpp:116]: ah thanks, I was looking for the right config. I'll clean this up

**Review by nicolasvasilache — COMMENTED** [review]
Hmm a bunch of stuff here is not GPU-specific though.
will move back to Common afterwards ..

File: compiler/src/iree/compiler/Codegen/LLVMGPU/GPUTransformDialectStrategies.cpp [hunk 2]
```diff
@@ -0,0 +1,423 @@
...
+
+  // clang-format off
+  tfdWithTileSizes<TileToForeachThreadOp>(b,
+                   /*rootH=*/parRedH,
+                   /*opsHToFuse=*/{fill1dH},
+                  // TODO: activate this, we have the information but it does
+                  // not generate the IR we want atm.
+                  //  /*tileSizes=*/tileSizes1Fill,
+                  //  /*threadDimMapping=*/b.getArrayAttr({y}));
+                   /*tileSizes=*/ArrayRef<int64_t>{1, 0, 0},
```
Comment (nicolasvasilache) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/GPUTransformDialectStrategies.cpp:283]: what's up with this?

File: compiler/src/iree/compiler/Codegen/LLVMGPU/GPUTransformDialectStrategies.cpp [hunk 3]
```diff
@@ -0,0 +1,423 @@
...
+                   /*threadDimMapping=*/b.getArrayAttr({z}));
+  tfdWithTileSizes<TileToForeachThreadOp>(b,
+                   /*rootH=*/parParRedH,
+                   /*opsHToFuse=*/{fill2dH},
+                  // TODO: activate this, we have the information but it does
+                  // not generate the IR we want atm.
+                  //  /*tileSizes=*/tileSizes1Generic,
+                  //  /*threadDimMapping=*/b.getArrayAttr({y}));
+                   /*tileSizes=*/ArrayRef<int64_t>{1, 1, 0},
```
Comment (nicolasvasilache) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/GPUTransformDialectStrategies.cpp:292]: what's up with this?

File: compiler/src/iree/compiler/Codegen/LLVMGPU/GPUTransformDialectStrategies.cpp [hunk 4]
```diff
@@ -0,0 +1,423 @@
...
+  Value originalGenericH =
+      b.create<MatchOp>(variantH, linalg::GenericOp::getOperationName());
+
+  // Step 1: Distribute to blocks using the current IREE lowering config.
+  variantH = buildReductionStrategyBlockDistributionPart(
+      b, variantH, originalFillH, originalGenericH, tileSizes[0]);
+
+  // Step 2. Second level of tiling + fusion parallelizes to threads.
+  variantH = buildReductionStrategyThreadDistributionPart(
+      b, variantH, tileSizes[1], tileSizes[1]);
```
Comment (nicolasvasilache) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/GPUTransformDialectStrategies.cpp:315]: this is unused at the client

File: compiler/src/iree/compiler/Codegen/LLVMGPU/GPUTransformDialectStrategies.cpp (hunk 2 above)
Comment (ThomasRaoux) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/GPUTransformDialectStrategies.cpp:283]: this is copy pasted, I have just moved code I didn't improve the matcher yet

File: compiler/src/iree/compiler/Codegen/LLVMGPU/GPUTransformDialectStrategies.cpp (hunk 3 above)
Comment (ThomasRaoux) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/GPUTransformDialectStrategies.cpp:292]: same.

**Pull request merged**